jax/core.py

```diff
@@ -995,48 +998,102 @@ def call_impl(f: lu.WrappedFun, *args, **params):
+  except TypeError:
+    return False
 
-  def read_env(env: Set[Var], v: Var):
-    if type(v) is not Literal and v not in env:
-      raise Exception("Variable '{}' not defined".format(v) + context())
+def typematch(aval1, aval2):
+  return (raise_to_shaped(aval1).strip_weak_type() ==
+          raise_to_shaped(aval2).strip_weak_type())
+
+class _JaxprContext(object):
+  __slots__ = ["env", "jaxpr"]
+  def __init__(self, jaxpr: Jaxpr):
+    self.env: Dict[Var, AbstractValue] = {}
+    self.jaxpr = jaxpr
+
+  def _msg_context(self):
+    return "\njaxpr:\n{}\n".format(self.jaxpr)
 
-  def write_env(env: Set[Var], v: Var):
+  def read_env(self, v: Var):
+    env = self.env
+    if type(v) is not Literal and v not in env:
+      raise Exception(
+          "Variable '{}' not defined".format(v) + self._msg_context())
+    if v.aval != env[v]:
+      raise TypeError(
+          "Variable '{}' inconsistently typed as {}, bound as {}".format(
+              v, v.aval, env[v]) + self._msg_context())
+    return v
+
+  def write_env(self, v: Var):
+    env = self.env
     if v in env:
-      raise Exception("Variable {} already bound".format(v) + context())
-    env.add(v)
+      raise Exception(
+          "Variable {} already bound".format(v) + self._msg_context())
+    env[v] = v.aval
+    return v
 
-  env: Set[Var] = set()
-  read = partial(read_env, env)
-  write = partial(write_env, env)
+def check_jaxpr(jaxpr: Jaxpr):
+  """Checks well-formedness of a jaxpr.
+
+  Specifically, check that:
+  - variables that are read are bound beforehand
+  - variables are typed equally throughout a jaxpr
+  - variable type annotations are compatible with their binding expression
+  """
+  ctx = _JaxprContext(jaxpr)
+  read = ctx.read_env
+  write = ctx.write_env
 
   write(unitvar)
   map(write, jaxpr.constvars)
   map(write, jaxpr.invars)
+
   for eqn in jaxpr.eqns:
-    if eqn.primitive.call_primitive:
-      if "call_jaxpr" not in eqn.params:
-        raise Exception("Call primitive {} should have a 'call_jaxpr' parameter"
-                        .format(eqn.primitive))
-    map(read, eqn.invars)
-    map(write, eqn.outvars)
+    check_jaxpr_eqn(ctx, eqn)
 
   for subjaxpr in subjaxprs(jaxpr):
     check_jaxpr(subjaxpr)
 
   map(read, jaxpr.outvars)
 
+def check_jaxpr_eqn(ctx, eqn):
+  prim = eqn.primitive
+
+  if prim.call_primitive:
+    if "call_jaxpr" not in eqn.params:
+      raise Exception("Call primitive {} should have a 'call_jaxpr' parameter"
+                      .format(prim))
+
+  invars = map(ctx.read_env, eqn.invars)
+  outvars = map(ctx.write_env, eqn.outvars)
+
+  in_avals = [v.aval for v in invars]
+  inferred_out_avals = prim.abstract_eval(*in_avals, **eqn.params)
+  if not prim.multiple_results:
+    inferred_out_avals = [inferred_out_avals]
+
+  for outvar, inferred_out_aval in zip(outvars, inferred_out_avals):
+    if not typecompat(outvar.aval, inferred_out_aval):
```

Comment:
Should this be `typecompat` or `typematch`?

Reply:
That's up to us to decide: what should be valid in the Jaxpr IR? I went with `typecompat` for now since it's the more relaxed option, but we can consider requiring `typematch` if we find the restriction useful, or if we find `typecompat` problematic. With `typecompat` we can, for example, infer a concrete array as the type of a Jaxpr equation's RHS, but still allow that it be bound to a variable (LHS) compatibly typed as a more abstract shaped array instead.